Issue #2069 Implemented tests for FolderObserver

m-memova opened the pull request:
There are 8 test cases testing the functionality of
org.eclipse.smarthome.model.core.internal.folder.FolderObserver.
They check if the correct ModelRepository's methods are invoked
when certain events in the watched directory are triggered.

Signed-off-by: Mihaela Memova <mihaela.memova@musala.com>

sjka commented:
Thanks @m-memova.

When I run those test locally on MacOS, I constantly get the following errors:

```
-------------------------------------------------------
 T E S T S
-------------------------------------------------------
Running org.eclipse.smarthome.model.core.internal.folder.test.FolderObserverTest
Tests run: 8, Failures: 2, Errors: 0, Skipped: 0, Time elapsed: 7.866 sec <<< FAILURE! - in org.eclipse.smarthome.model.core.internal.folder.test.FolderObserverTest
test the updating of the configuration with a non existing subdirectory(org.eclipse.smarthome.model.core.internal.folder.test.FolderObserverTest)  Time elapsed: 0.4 sec
test modification of file with valid extension in an existing subdirectory(org.eclipse.smarthome.model.core.internal.folder.test.FolderObserverTest)  Time elapsed: 1.625 sec  <<< FAILURE!
java.lang.AssertionError: A model was not added/refreshed adequately on valid file modification in the watched directory
Expected: is <true>
     but: was <false>
       	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
       	at org.junit.Assert.assertThat(Assert.java:865)
       	at org.junit.Assert$assertThat$0.callStatic(Unknown Source)
       	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCallStatic(CallSiteArray.java:53)
       	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callStatic(AbstractCallSite.java:157)
       	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callStatic(AbstractCallSite.java:173)
       	at org.eclipse.smarthome.model.core.internal.folder.test.FolderObserverTest.test modification of file with valid extension in an existing subdirectory(FolderObserverTest.groovy:232)

test deletion of a folder-extensions pair from the configuration properties when folderFileExtMap remains non-empty(org.eclipse.smarthome.model.core.internal.folder.test.FolderObserverTest)  Time elapsed: 1.043 sec
test deletion of a folder-extensions pair from the configuration properties when folderFileExtMap becomes empty(org.eclipse.smarthome.model.core.internal.folder.test.FolderObserverTest)  Time elapsed: 0.422 sec
test creation of file with valid extension in an existing subdirectory(org.eclipse.smarthome.model.core.internal.folder.test.FolderObserverTest)  Time elapsed: 1.034 sec  <<< FAILURE!
java.lang.AssertionError: A model was not added/refreshed adequately on new valid file creation in the watched directory
Expected: is <true>
     but: was <false>
       	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
       	at org.junit.Assert.assertThat(Assert.java:865)
       	at org.junit.Assert$assertThat$0.callStatic(Unknown Source)
       	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCallStatic(CallSiteArray.java:53)
       	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callStatic(AbstractCallSite.java:157)
       	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callStatic(AbstractCallSite.java:173)
       	at org.eclipse.smarthome.model.core.internal.folder.test.FolderObserverTest.test creation of file with valid extension in an existing subdirectory(FolderObserverTest.groovy:186)

test creation of file with extension which is not in the configuration properties in an existing subdirectory(org.eclipse.smarthome.model.core.internal.folder.test.FolderObserverTest)  Time elapsed: 1.016 sec
test creation of file in subdirectory with no declared extensions(org.eclipse.smarthome.model.core.internal.folder.test.FolderObserverTest)  Time elapsed: 1.019 sec
test creation of file with extension which does not have a registered ModelParser in an existing subdirectory(org.eclipse.smarthome.model.core.internal.folder.test.FolderObserverTest)  Time elapsed: 1.022 sec

Results :

Failed tests:
  FolderObserverTest.test modification of file with valid extension in an existing subdirectory:232 A model was not added/refreshed adequately on valid file modification in the watched directory
Expected: is <true>
     but: was <false>
  FolderObserverTest.test creation of file with valid extension in an existing subdirectory:186 A model was not added/refreshed adequately on new valid file creation in the watched directory
Expected: is <true>
     but: was <false>

Tests run: 8, Failures: 2, Errors: 0, Skipped: 0
```
Could you please have a look? If that is something that might behave differently depending on the platform, let me know if I should try out anything for you.

m-memova commented:
Thank you, @sjka . I will take a look.

m-memova commented:
I try to update this PR with my fixed tests but the Travis CI build failed with : 
> [INFO] Eclipse SmartHome Automation Integration Tests .... FAILURE [20.503s]
I am not sure what should I do in that case.
@sjka can you please check the tests on your machine?

maggu2810 commented:
@sjka Can you test this again?

maggu2810 commented on bundles/model/org.eclipse.smarthome.model.core.test/pom.xml:
```diff
@@ -0,0 +1,82 @@
+    </plugins>
+  </build>
+
+</project>
```
IIRC we are using 2 spaces for indent of pom files. It seems you are mixing tabs and spaces in that file and the indentation is not correct on all places (e.g. no indentation on the starting build element).

maggu2810 commented on bundles/model/org.eclipse.smarthome.model.core.test/src/test/groovy/org/eclipse/smarthome/model/core/internal/folder/test/FolderObserverTest.groovy:
```diff
@@ -0,0 +1,535 @@
+		config.delete()
+	}
+
+	void deleteDirectory(File path) {
```
Okay for me, but isn't the Apache Commons IO artifact already available?
There is a static method `FileUtils.deleteDirectory(File directory)`

maggu2810 commented:
Tested it on my local machine and the tests succeeded.
```text
Tests run: 8, Failures: 0, Errors: 0, Skipped: 0
```

m-memova commented:
Thank you, @maggu2810, I fixed the problems, you mentioned and I updated the PR. In addition, I tried to run the tests on a Mac machine and everything seems fine there.
